torch/csrc/api/src/optim/adam.cpp

```diff
@@ -12,57 +12,171 @@
+         (lhs.beta2() == rhs.beta2()) &&
+         (lhs.eps() == rhs.eps()) &&
+         (lhs.weight_decay() == rhs.weight_decay() &&
+         (lhs.amsgrad() == rhs.amsgrad()));
+}
+
+void AdamOptions::serialize(torch::serialize::OutputArchive& archive) const {
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(lr);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(beta1);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(beta2);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(eps);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(weight_decay);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(amsgrad);
+}
+
+void AdamOptions::serialize(torch::serialize::InputArchive& archive) {
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(double, lr);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(double, beta1);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(double, beta2);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(double, eps);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(double, weight_decay);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(bool, amsgrad);
+}
+
+bool operator==(const AdamParamState& lhs, const AdamParamState& rhs) {
+  return (lhs.step() == rhs.step()) &&
+          torch::equal(lhs.exp_avg(), rhs.exp_avg()) &&
+          torch::equal(lhs.exp_avg_sq(), rhs.exp_avg_sq()) &&
+          !(lhs.max_exp_avg_sq().defined() ^ rhs.max_exp_avg_sq().defined());
+}
+
+void AdamParamState::serialize(torch::serialize::OutputArchive& archive) const {
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(step);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(exp_avg);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(exp_avg_sq);
+  if(max_exp_avg_sq().defined()) {
+    _TORCH_OPTIM_SERIALIZE_TORCH_ARG(max_exp_avg_sq);
+  }
+}
+
+void AdamParamState::serialize(torch::serialize::InputArchive& archive) {
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(int64_t, step);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(Tensor, exp_avg);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(Tensor, exp_avg_sq);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(Tensor, max_exp_avg_sq);
+}
 
 void Adam::step() {
-  for (size_t i = 0; i < parameters_.size(); ++i) {
-    Tensor p = parameters_.at(i);
-    if (!p.grad().defined()) {
-      continue;
-    }
+  for (auto& group : param_groups_) {
+    for (auto& p : group.params()) {
+      if (!p.grad().defined()) {
+        continue;
+      }
+      auto grad = p.grad().data();
+      TORCH_CHECK(!grad.is_sparse(), "Adam does not support sparse gradients"/*, please consider SparseAdam instead*/);
+      auto param_state = state_.find(c10::guts::to_string(p.unsafeGetTensorImpl()));
+      auto& options = static_cast<AdamOptions&>(group.options());
+      // State initialization
+      if(param_state == state_.end()) {
+        auto state = std::make_unique<AdamParamState>();
+        state->step(0);
+        state->exp_avg(torch::zeros_like(p.data(), MemoryFormat::Preserve));
+        state->exp_avg_sq(torch::zeros_like(p.data(), MemoryFormat::Preserve));
+        if(options.amsgrad()) {
+          state->max_exp_avg_sq(torch::zeros_like(p.data(), MemoryFormat::Preserve));
+        } else {
+          state->max_exp_avg_sq({});
```

Comment:
Since we are setting default value of `max_exp_avg_sq` to undefined tensor, we probably don't need to explicitly set its value to undefined tensor here.